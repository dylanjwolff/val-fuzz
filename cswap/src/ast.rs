use crate::Timer;
use std::cell::RefCell;
use std::fmt;
use std::fs;
use std::path::Path;
use std::rc::Rc;
use std::time::Duration;
#[macro_use]
use serde::{Serialize, Deserialize};

pub type ScriptRc = Rc<RefCell<Script>>;
pub type CommandRc = Rc<RefCell<Command>>;
pub type LogicRc = Rc<RefCell<Logic>>;
pub type SExpRc = Rc<RefCell<SExp>>;
pub type SExpBoxRc = Rc<RefCell<Box<SExp>>>;
pub type SortRc = Rc<RefCell<Sort>>;
pub type SymbolRc = Rc<RefCell<Symbol>>;
pub type ConstantRc = Rc<RefCell<Constant>>;
pub type BoolOpRc = Rc<RefCell<BoolOp>>;

macro_rules! rccell {
    ($x:expr) => {{
        std::rc::Rc::new(std::cell::RefCell::new($x))
    }};
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Script {
    Commands(Vec<CommandRc>),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Command {
    Logic(LogicRc),
    CheckSat(),
    CheckSatAssuming(SExpRc),
    Assert(SExpRc),
    GetModel(),
    DeclConst(String, SortRc),
    Generic(String),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Sort {
    UInt(),
    Dec(),
    Str(),
    Bool(),
    BitVec(u32),
    Array(),
    UserDef(String),
    Compound(Vec<SortRc>),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum SExp {
    Compound(Vec<SExpRc>),
    Let(Vec<(SymbolRc, SExpRc)>, SExpBoxRc),
    QForAll(Vec<(SymbolRc, SortRc)>, SExpBoxRc),
    BExp(BoolOpRc, Vec<SExpRc>),
    Constant(ConstantRc),
    Symbol(SymbolRc),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Symbol {
    Var(String), // Currently, it is hard to detect all variables, so some Tokens may also be
    // variables too
    Token(String),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum BoolOp {
    Equals(),
    And(),
    Or(),
    Xor(),
    Implies(),
    Distinct(),
    Gt(),
    Lt(),
    Gte(),
    Lte(),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Constant {
    UInt(String),
    Dec(String),
    Hex(String),
    Bin(Vec<char>),
    Str(String),
    Bool(bool),
}

#[allow(non_camel_case_types)]
#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum Logic {
    QF_SLIA(),
    Other(String),
}

#[derive(Serialize, Deserialize, Debug, Eq, PartialEq, Clone)]
pub enum AstNode {
    Script(ScriptRc),
    Command(CommandRc),
    Constant(ConstantRc),
    Symbol(SymbolRc),
    SExp(SExpRc),
    Logic(LogicRc),
    BoolOp(BoolOpRc),
    Sort(SortRc),
    Open(),
    Close(),
}

impl Script {
    pub fn to_f(&mut self, filepath: &Path) {
        fs::write(filepath, self.to_string());
    }

    pub fn insert(&mut self, i: usize, cmd: Command) {
        let Script::Commands(cmds) = self;
        if i > cmds.len() {
            cmds.push(rccell!(cmd));
        } else {
            cmds.insert(i, rccell!(cmd));
        }
    }

    pub fn replace(&mut self, i: usize, cmd: Command) {
        let Script::Commands(cmds) = self;
        cmds[i] = rccell!(cmd);
    }

    pub fn init(&mut self, i: usize) {
        let Script::Commands(cmds) = self;
        cmds.insert(i, rccell!(Command::Assert(rccell!(SExp::true_sexp()))));
    }

    pub fn index_is_gm(&self, i: usize) -> bool {
        let Script::Commands(cmds) = self;
        if i >= cmds.len() {
            return false;
        }
        match &*cmds[i].borrow() {
            Command::GetModel() => true,
            _ => false,
        }
    }

    pub fn is_unsupported_logic(&self) -> bool {
        let Script::Commands(cmds) = self;
        let maybe_logic = cmds.iter().find(|cmd| cmd.borrow().is_logic());
        match maybe_logic {
            Some(cmd) => match &*cmd.borrow() {
                Command::Logic(l) => match *l.borrow() {
                    Logic::QF_SLIA() => true,
                    _ => false,
                },
                _ => false,
            },
            _ => false,
        }
    }
}

impl fmt::Display for Script {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Script::Commands(cmds) => cmds
                .iter()
                .map(|cmd| write!(f, "{}\n", cmd.borrow()))
                .fold(Ok(()), acc_result),
        }
    }
}

impl Command {
    pub fn is_logic(&self) -> bool {
        match self {
            Command::Logic(_) => true,
            _ => false,
        }
    }

    pub fn is_checksat(&self) -> bool {
        match self {
            Command::CheckSat() | Command::CheckSatAssuming(_) => true,
            _ => false,
        }
    }

    pub fn no_op() -> Self {
        Command::Generic("".to_string())
    }
}

impl fmt::Display for Command {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Command::Logic(l) => write!(f, "(set-logic {})", l.borrow()),
            Command::CheckSat() => write!(f, "(check-sat)"),
            Command::CheckSatAssuming(sexp) => write!(f, "(check-sat-assuming {})", sexp.borrow()),
            Command::GetModel() => write!(f, "(get-model)"),
            Command::DeclConst(v, s) => write!(f, "(declare-const {} {})", v, s.borrow()),
            Command::Generic(s) => write!(f, "{}", s),
            Command::Assert(s) => write!(f, "(assert {})", s.borrow()),
        }
    }
}

impl fmt::Display for Constant {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Constant::UInt(s) => write!(f, "{}", s),
            Constant::Dec(d) => write!(f, "{}", d),
            Constant::Hex(s) => write!(f, "#x{}", s),
            Constant::Str(s) => write!(f, "\"{}\"", s),
            Constant::Bool(b) => write!(f, "{}", b),
            Constant::Bin(bv) => write!(f, "#b{}", bv.into_iter().collect::<String>()),
        }
    }
}

fn acc_result(acc: fmt::Result, curr: fmt::Result) -> fmt::Result {
    match curr {
        Ok(_) => acc,
        Err(e) => Err(e),
    }
}

impl fmt::Display for Sort {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Sort::UInt() => write!(f, "Int"),
            Sort::Dec() => write!(f, "Real"),
            Sort::Bool() => write!(f, "Bool"),
            Sort::Str() => write!(f, "String"),
            Sort::BitVec(len) => write!(f, "(_ BitVec {})", len),
            Sort::Array() => write!(f, "Array"),
            Sort::UserDef(s) => write!(f, "{}", s),
            Sort::Compound(v) => {
                write!(f, "(")?;
                v.iter()
                    .enumerate()
                    .map(|(i, sort)| match i == 0 {
                        true => write!(f, "{}", sort.borrow()),
                        false => write!(f, " {}", sort.borrow()),
                    })
                    .fold(Ok(()), acc_result)?;
                write!(f, ")")
            }
        }
    }
}

impl SExp {
    pub fn true_sexp() -> SExp {
        SExp::Symbol(rccell!(Symbol::Token("true".to_owned())))
    }

    pub fn false_sexp() -> SExp {
        SExp::Symbol(rccell!(Symbol::Token("false".to_owned())))
    }
}

impl fmt::Display for SExp {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            SExp::Constant(c) => write!(f, "{}", c.borrow()),
            SExp::Symbol(s) => write!(f, "{}", s.borrow()),
            SExp::Let(vbs, s) => {
                write!(f, "(let (")?;
                vbs.iter()
                    .map(|(v, s)| write!(f, "({} {})", v.borrow(), s.borrow()))
                    .fold(Ok(()), acc_result)?;
                write!(f, " {})", s.borrow())
            }
            SExp::Compound(v) => {
                write!(f, "(")?;
                v.iter()
                    .enumerate()
                    .map(|(i, sexp)| match i == 0 {
                        true => write!(f, "{}", sexp.borrow()),
                        false => write!(f, " {}", sexp.borrow()),
                    })
                    .fold(Ok(()), acc_result)?;
                write!(f, ")")
            }
            SExp::BExp(o, v) => {
                write!(f, "({} ", o.borrow())?;
                v.iter()
                    .enumerate()
                    .map(|(i, sexp)| match i == 0 {
                        true => write!(f, "{}", sexp.borrow()),
                        false => write!(f, " {}", sexp.borrow()),
                    })
                    .fold(Ok(()), acc_result)?;
                write!(f, ")")
            }
            SExp::QForAll(v, s) => {
                write!(f, "(forall (")?;
                v.iter()
                    .map(|(va, vl)| write!(f, "({} {})", va.borrow(), vl.borrow()))
                    .fold(Ok(()), acc_result)?;
                write!(f, ") {})", s.borrow())
            }
        }
    }
}

impl fmt::Display for Symbol {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Symbol::Var(s) | Symbol::Token(s) => write!(f, "{}", s),
        }
    }
}

impl fmt::Display for BoolOp {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            BoolOp::And() => write!(f, "and"),
            BoolOp::Or() => write!(f, "or"),
            BoolOp::Xor() => write!(f, "xor"),
            BoolOp::Implies() => write!(f, "=>"),
            BoolOp::Distinct() => write!(f, "distinct"),
            BoolOp::Equals() => write!(f, "="),
            BoolOp::Gt() => write!(f, ">"),
            BoolOp::Lt() => write!(f, "<"),
            BoolOp::Gte() => write!(f, ">="),
            BoolOp::Lte() => write!(f, "<="),
        }
    }
}

impl fmt::Display for Logic {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Logic::QF_SLIA() => write!(f, "QF_SLIA"),
            Logic::Other(s) => write!(f, "{}", s),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::*;
    use insta::assert_debug_snapshot;
    use walkdir::WalkDir;

    #[test]
    fn sort_display_snap() {
        let sort = Sort::Compound(vec![rccell!(Sort::UInt()), rccell!(Sort::Bool())]);
        assert_debug_snapshot!(format!("{}", sort));
    }

    #[test]
    fn let_to_string_snap() {
        assert_debug_snapshot!(script(
            "(assert (let ((max (seq.nth A m))(n   (seq.len A))) (ite (or true false))))"
        )
        .unwrap()
        .1
        .to_string());
    }
}
