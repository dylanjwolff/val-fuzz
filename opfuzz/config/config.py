solvers = [
]

crash_msgs = [
    "Exception",
    "lang.AssertionError",
    "lang.Error",
    "runtime error",
    "LEAKED",
    "Leaked",
    "Segmentation fault",
    "segmentation fault",
    "segfault",
    "ASSERTION",
    "Assertion",
    "Fatal failure",
    "Internal error detected",
    "an invalid model was generated",
    "Failed to verify",
    "failed to verify",
    "ERROR: AddressSanitizer:",
    "invalid expression",
    "Aborted"
]

ignore_msgs = [
    "unexpected char",
    "failed to open file",
    "Expected result sat but got unsat",
    "Expected result unsat but got sat",
    "Parse Error",
    "Cannot get model",
    "Symbol 'str.to-re' not declared as a variable",
    "Symbol 'str.to.re' not declared as a variable",
    "Unimplemented code encountered",
]
