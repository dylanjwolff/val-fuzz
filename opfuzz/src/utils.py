import random
import string

def random_string(length=5):
    return ''.join(random.sample(string.ascii_letters + string.digits, length))

def plain(cli):
    plain_cli = ""
    for token in cli.split(" "):
        plain_cli += token.split("/")[-1]
    return escape(plain_cli)

def escape(s):
    s = s.replace(".", "")
    s = s.replace("=", "")
    return s

