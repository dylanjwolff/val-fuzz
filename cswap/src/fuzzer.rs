use crate::ast::CommandRc;
use crate::ast::Script;
use crate::ast::Sort;
use crate::config::Config;
use crate::config::FileProvider;
use crate::config::Metadata;
use crate::parser::script_from_f;
use crate::parser::script_from_f_unsanitized;
use crate::solver::check_valid_solve;
use crate::solver::profiles_solve;
use crate::solver::randomized_profiles_solve;
use crate::solver::ProfileIndex;
use crate::solver::RSolve;
use crate::transforms::rv;
use crate::utils::dyn_fmt;
use crate::utils::to_strs;
use crate::utils::DFormatParseError;
use crate::utils::HashHashSet;
use crate::utils::RunStats;
use std::collections::HashMap;

use crate::utils::RandUniqPermGen;
use std::hash::Hash;

use std::collections::BTreeMap;
use std::fs::OpenOptions;
use std::io;
use walkdir::WalkDir;

use crate::ast::BoolOp;
use crate::ast::SExp;
use crate::transforms::{
    ba_script, end_insert_pt, get_bav_assign_fmt_str, many_assert,
    replace_constants_with_fresh_vars,
};
use std::collections::HashSet;

use crate::liftio;
use crate::liftio_e;
use crate::solver::all_non_err_timed_out;
use bit_vec::BitVec;
use log::debug;
use log::trace;
use log::warn;

use std::fs;
use std::io::Write;
use std::path::Path;
use std::path::PathBuf;

pub fn exec_single_thread(dirname: &str, cfg: Config) {
    let mut q = vec![];
    for entry in WalkDir::new(dirname)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|e| !e.file_type().is_dir())
    {
        let filepath = entry.into_path();
        debug!("push {:?}", filepath);
        q.push(filepath);
    }

    let mut stats = RunStats::new();
    let mut to_ctr = BTreeMap::new();
    for f in q.into_iter() {
        match mutator(f.clone(), &cfg)
            .map(|skels| {
                skels
                    .into_iter()
                    .flat_map(|skel| single_thread_group_bav_assign(skel, &cfg))
                    .flat_map(|skelfs| skelfs)
                    .collect::<Vec<(PathBuf, PathBuf)>>()
            })
            .map(|iters| {
                for iter in iters.into_iter() {
                    match solver_fn(iter, &mut to_ctr, &mut stats, vec![], &cfg) {
                        Err(e) => warn!("Solver Error: {} in file {:?}", e, f),
                        _ => (),
                    };
                }
            }) {
            Err(e) => warn!("{} in file {:?}", e, f),
            _ => (),
        };
    }
}

pub fn mutator(filepath: PathBuf, cfg: &Config) -> io::Result<Vec<(PathBuf, PathBuf)>> {
    let (scripts, mut md) = strip_and_transform(&filepath, cfg)?;

    let mut mutatedfs = vec![];
    for script in scripts {
        let (skelf, mdf) = cfg.file_provider.serialize_skel_and_md(&script, &mut md)?;
        mutatedfs.push((skelf.to_path_buf(), mdf.to_path_buf()));
    }
    Ok(mutatedfs)
}

pub struct StatefulBavAssign<'a> {
    top_of_script: Box<String>,
    bottom_of_script: Box<String>,
    pub md: Metadata,
    pub urng: RandUniqPermGen,
    cfg: &'a Config,
}

impl<'a> StatefulBavAssign<'a> {
    pub fn new(
        filepaths: (PathBuf, PathBuf),
        cfg: &'a Config,
    ) -> io::Result<StatefulBavAssign<'a>> {
        let (script, md) = deserialize_from_f(&filepaths)?;
        trace!("Validating skeleton for {}", md.seed_file);

        let e = liftio_e!(Err::<(), &str>("Filename not a str"));
        let empty_skel_fstr = &filepaths.0.to_str().ok_or(e)?;

        let results = check_valid_solve(empty_skel_fstr);

        report_any_bugs(&filepaths.0, &results, &cfg.file_provider);

        if results
            .iter()
            .all(|r| r.has_unrecoverable_error() && !r.has_bug_error())
        {
            if cfg.remove_files {
                fs::remove_file(&filepaths.0).unwrap_or(());
                fs::remove_file(&filepaths.1).unwrap_or(());
            }

            let mut stresults = results
                .iter()
                .map(|o| o.to_string())
                .collect::<Vec<String>>()
                .join("\n----------\n");

            stresults = format!("============== \n{}\n =============", stresults);

            let errmsg = format!("All error on file {:?}", filepaths.0);
            warn!("{}\n{}", errmsg, stresults);
            return liftio!(Err(errmsg));
        }

        if results
            .iter()
            .filter(|r| !r.has_unrecoverable_error())
            .any(|r| r.has_unsat() && !r.has_sat() && !r.has_unknown())
        {
            return liftio!(Err("File has unsat skeleton!"));
        }

        let (top, bottom) = Self::split(script);
        let num_bavs = md.bavns.len();

        let urng = RandUniqPermGen::new_masked_with_retries(
            cfg.get_specific_seed(&md.seed_file),
            num_bavs,
            cfg.max_iter,
            cfg.mask_size,
        );

        Ok(StatefulBavAssign {
            top_of_script: top,
            bottom_of_script: bottom,
            md: md,
            urng: urng,
            cfg: cfg,
        })
    }

    fn split(script: Script) -> (Box<String>, Box<String>) {
        let eip = end_insert_pt(&script);
        let (top, bottom) = Script::split(script, eip);
        (Box::new(top.to_string()), Box::new(bottom.to_string()))
    }

    pub fn do_iteration_tv(
        &mut self,
        i: u64,
        truth_values: BitVec,
        mask: BitVec,
    ) -> io::Result<(PathBuf, PathBuf)> {
        let new_file: PathBuf = self.cfg.file_provider.iterfile(i, &self.md)?;
        let mut f = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&new_file)?;

        let bav_unenforced = get_bav_assign_fmt_str(&self.md.bavns);
        let bav_uef_filtered = bav_unenforced
            .into_iter()
            .zip(mask.iter())
            .filter_map(|(cmd, mbit)| if mbit { Some(cmd) } else { None })
            .collect::<Vec<CommandRc>>();

        let num_remaining = bav_uef_filtered.len();
        let bav_fmt_string = Script::Commands(bav_uef_filtered).to_string();

        assert!(
            num_remaining == truth_values.len(),
            format!(
                "fmt str size {:?} != {:?} num truth vals from mask {:?}",
                bav_fmt_string,
                truth_values.len(),
                mask,
            )
        );
        assert!(
            mask.iter().fold(0, |a, b| if b { a + 1 } else { a }) == truth_values.len(),
            "Mask size != num truth vals"
        );
        let str_with_model =
            liftio!(dyn_fmt(&bav_fmt_string, to_strs(&truth_values))
                .map_err(|e| slim_dynfmt_err_msg(e)))?;

        f.write_all(self.top_of_script.as_bytes())?;
        f.write_all(str_with_model.as_bytes())?;
        f.write_all(self.bottom_of_script.as_bytes())?;

        Ok((
            new_file.to_path_buf(),
            self.md.md_path(&self.cfg.file_provider),
        ))
    }
}

fn single_thread_group_bav_assign(
    filepaths: (PathBuf, PathBuf),
    cfg: &Config,
) -> io::Result<Vec<(PathBuf, PathBuf)>> {
    let mut sba = StatefulBavAssign::new(filepaths, cfg)?;
    let mut fs = vec![];
    let mut i = 0;

    while let Some((tv, mask)) = sba.urng.sample_and_mask() {
        let _ef = sba
            .md
            .bavns
            .iter()
            .zip(mask.iter())
            .filter_map(|((name, _sort), mbit)| if mbit { Some(name) } else { None })
            .zip(tv.iter())
            .collect::<Vec<(&String, bool)>>();

        fs.push(sba.do_iteration_tv(i, tv, mask)?);
        i = i + 1;
    }
    Ok(fs)
}
fn slim_dynfmt_err_msg(e: DFormatParseError) -> String {
    match e {
        nom::Err::Incomplete(n) => format!("Incomplete parse on Dynamic Fmt; needed {:?}", n),
        nom::Err::Error(ec) => format!("Dynamic format parse error of kind {:?}", ec.1),
        nom::Err::Failure(ec) => format!("Dynamic format parse failure of kind {:?}", ec.1),
    }
}

pub fn report_any_bugs(file: &Path, results: &Vec<RSolve>, fp: &FileProvider) -> bool {
    results
        .iter()
        .find(|r| r.has_bug_error())
        .map(|r| {
            debug!("Reporting bug for file {:?}", file);
            fp.bug_report(file, &format!("{}", r));
        })
        .is_some()
        || if !RSolve::differential_test(&results).is_ok() {
            debug!("Reporting soundness bug for file {:?}", file);
            fp.bug_report(file, &format!("SOUNDNESS BUG: {:?}", results));
            true
        } else {
            false
        }
}

fn deserialize_from_f(
    (script_file, md_file): &(PathBuf, PathBuf),
) -> io::Result<(Script, Metadata)> {
    trace!("Reading script from {:?}", script_file);
    let script = script_from_f_unsanitized(script_file)?;

    trace!("Reading md from {:?}", md_file);
    let md_contents = fs::read_to_string(&md_file)?;
    let md: Metadata = liftio!(serde_lexpr::from_str(&md_contents))?;

    Ok((script, md))
}

lazy_static! {
    static ref SIMPLE_PROFILE: HashSet<ProfileIndex> =
        vec![ProfileIndex::Z3(0), ProfileIndex::CVC4(1)]
            .into_iter()
            .collect();
}
pub fn solver_fn(
    filepaths: (PathBuf, PathBuf),
    to_ctr: &mut BTreeMap<String, u64>,
    stats: &mut RunStats,
    enforcement: Vec<(String, bool)>,
    cfg: &Config,
) -> io::Result<()> {
    let (iter_file, md_file) = &filepaths;
    let mdstr = fs::read_to_string(md_file)
        .map_err(|e| io::Error::new(io::ErrorKind::Other, format!("File {:?} {:?}", md_file, e)))?;
    let md: Metadata = liftio!(serde_lexpr::from_str(&mdstr))?;
    if to_ctr.get(&md.seed_file).map(|sf| *sf > 3).unwrap_or(false) {
        fs::remove_file(md_file).unwrap_or(()); // TODO Hack! Prevents other threads from using the file
        return liftio!(Err(format!("{:?} exceeded max consec timeouts", iter_file)));
    }

    let seed = cfg.get_specific_seed(iter_file);
    let results = randomized_profiles_solve(
        filepaths.0.to_str().unwrap_or("defaultname"),
        &SIMPLE_PROFILE,
        seed,
        cfg.timeout,
    );

    let ct = to_ctr.entry(md.seed_file.clone()).or_insert(0);
    if all_non_err_timed_out(&results) {
        *ct = *ct + 1;
    } else if *ct > 0 {
        *ct = *ct - 1;
    }

    stats.record_iter();
    stats.record_stats_for_iter_results(&results);

    results.iter().filter(|r| r.has_sat()).for_each(|r| {
        match resub_model(r, &filepaths, &md, stats, &enforcement, &cfg) {
            Ok(()) => (),
            Err(e) => warn!("Resub error {} for file {:?}", e, iter_file),
        }
    });

    if !report_any_bugs(filepaths.0.as_path(), &results, &cfg.file_provider) {
        if cfg.remove_files {
            fs::remove_file(iter_file).unwrap_or(());
        }
    }
    Ok(())
}

fn resub_model(
    result: &RSolve,
    filepaths: &(PathBuf, PathBuf),
    md: &Metadata,
    stats: &mut RunStats,
    enforcemt: &Vec<(String, bool)>,
    cfg: &Config,
) -> io::Result<()> {
    debug!("RESUB! for file {:?}", filepaths.0);

    let f_to_replace = match md.ogwms_path(&cfg.file_provider) {
        Some(ogwms_f) => ogwms_f,
        None => md.choles_path(&cfg.file_provider),
    };

    let mut choles_script = script_from_f_unsanitized(&f_to_replace)?;
    // add enforcement here
    if cfg.enforce_on_resub {
        assert!(
            cfg.monitors_in_final,
            "Must have monitors in final to enforce resub!"
        );
        let efcmt_cmds = many_assert(&mut enforcemt.iter().map(|(name, val)| {
            SExp::BExp(
                rccell!(BoolOp::Equals()),
                vec![rccell!(SExp::var(&name)), rccell!(SExp::bool_sexp(*val))],
            )
        }));
        choles_script.insert_all(end_insert_pt(&choles_script), &efcmt_cmds);
    }

    let mut to_replace: Vec<(String, SExp)> = result
        .extract_const_var_vals(&md.constvns.iter().map(|c| c.0.clone()).collect())
        .into_iter()
        .map(|(sym, val)| (sym.to_string(), val.clone()))
        .collect();

    if to_replace.len() > 0 {
        rv(&mut choles_script, &mut to_replace);

        let script_str = choles_script.to_string();
        if !stats.record_sub(&script_str) {
            return Ok(()); // Reporting this as an error is probably a bit too noisy
        }

        let resubbed_f = cfg
            .file_provider
            .serialize_resub_str(script_str, &filepaths.0, &md)?;

        let results = profiles_solve(
            resubbed_f.to_str().unwrap_or("defaultname"),
            &cfg.profiles,
            cfg.timeout,
        );
        stats.record_stats_for_sub_results(&results);

        log_check_enforce(&results, enforcemt);
        stats
            .coverage
            .log_check_coverage(&results, &md.bavns, md.seed_file.clone());

        if !report_any_bugs(&resubbed_f, &results, &cfg.file_provider) {
            if cfg.remove_files {
                fs::remove_file(&resubbed_f).unwrap_or(());
            }
        }
    }
    Ok(())
}

/// Exit on first success, only report if no successes and at least one failure
fn log_check_enforce(results: &Vec<RSolve>, enforcemt: &Vec<(String, bool)>) {
    let (enames, _evals): (Vec<_>, Vec<_>) = enforcemt.iter().cloned().unzip();

    let strress = results
        .iter()
        .flat_map(|result| {
            result
                .extract_const_var_vals(&enames)
                .into_iter()
                .map(|(sym, sexp)| (sym.to_string(), sexp.to_string()))
        })
        .collect::<Vec<(String, String)>>(); // TODO clean up (no string comp)

    enforcemt.iter().for_each(|(ename, eval)| {
        let filtered = strress
            .iter()
            .filter(|(name, _strval)| *name == *ename)
            .collect::<Vec<&(String, String)>>();
        if filtered.len() > 0 {
            match filtered
                .iter()
                .find(|(_, strval)| eval.to_string() == *strval)
            {
                Some(_) => trace!("SUCCESS ENFORCING {} to be {}", ename, eval),
                None => trace!("FAILURE ENFORCING {} to be {}", ename, eval),
            }
        }
    });
}

pub fn strip_and_transform(
    source_file: &Path,
    cfg: &Config,
) -> io::Result<(Vec<Script>, Metadata)> {
    let mut script = script_from_f(source_file)?;

    let og_f_results = check_valid_solve(&liftio!(source_file.to_str().ok_or(
        Err::<&str, String>(format!("Non-str Filename: {:?}!", source_file))
    ))?);

    if og_f_results
        .iter()
        .all(|r| r.has_unrecoverable_error() && !r.has_bug_error())
    {
        return liftio!(Err(format!(
            "Unmodified file {:?} failed to pass initial validation check",
            source_file
        )));
    }

    let mut md = Metadata::new(source_file);

    replace_constants_with_fresh_vars(&mut script, &mut md, cfg)?;
    let chf = cfg.file_provider.cholesfile(&mut md)?;
    fs::write(chf, script.to_string())?;

    let ba_script = ba_script(&mut script, &mut md, cfg)?;
    if cfg.monitors_in_final {
        cfg.file_provider.serialize_og_w_m(&script, &mut md)?;
    }

    return Ok((ba_script, md));
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::ast::Sort;
    use crate::parser::script;

    use bit_vec::BitVec;

    use insta::assert_display_snapshot;
    use std::collections::HashSet;
    use tempfile::TempDir;

    #[test]
    fn ba_script_num() {
        let mut script = script("(declare-fun z () Real)(assert (= 3 (+ z 4)))")
            .unwrap()
            .1;
        let mut md = Metadata::new_empty();
        let cfg = Config::default();
        replace_constants_with_fresh_vars(&mut script, &mut md, &cfg).unwrap();
        let new_script = &ba_script(
            &mut script,
            &mut md,
            &Config {
                use_bdom_vs: true,
                max_const_relations_to_monitor: 2,
                ..Config::default()
            },
        )
        .unwrap()[0];
        assert_display_snapshot!(new_script);
    }

    #[test]
    fn ba_script_fpop() {
        let mut script = script("(declare-const mpfx (_ FloatingPoint 11 53))(assert (= mpfx (fp.mul roundTowardPositive
            ((_ to_fp 11 53) roundNearestTiesToEven 0.5792861143265499723753464422770775854587554931640625 (- 1022))
            ((_ to_fp 11 53) roundNearestTiesToEven 1.3902774452208657152141313417814671993255615234375 (- 17)))))").unwrap().1;
        let mut md = Metadata::new_empty();
        let cfg = Config::default();
        replace_constants_with_fresh_vars(&mut script, &mut md, &cfg).unwrap();
        let new_script = &ba_script(
            &mut script,
            &mut md,
            &Config {
                use_bdom_vs: true,
                ..Config::default()
            },
        )
        .unwrap()[0];
        assert_display_snapshot!(new_script);
    }

    #[test]
    fn bav_stateful() {
        let script =
            script("(decl-const BAV1 bool)(assert (< x 5))(assert (= BAV1 (< x 5)))(check-sat)")
                .unwrap()
                .1;
        let bavns = vec![
            ("BAV1".to_string(), Sort::Bool()),
            ("BAV2".to_string(), Sort::Bool()),
        ];
        let (top, bottom) = StatefulBavAssign::split(script);
        let num_bavs = 2;
        let urng = RandUniqPermGen::new_definite_seeded(1, num_bavs, 1);
        let mut md = Metadata::new_empty();
        md.bavns = bavns;
        let base = TempDir::new().unwrap();
        let mut fpdir = base.path().to_path_buf();
        fpdir.push("test");

        let cfg = Config {
            file_provider: FileProvider::new(fpdir.to_str().unwrap()),
            max_iter: 1,
            rng_seed: 1,
            remove_files: true,
            ..Config::default()
        };

        let mut sba = StatefulBavAssign {
            top_of_script: top,
            bottom_of_script: bottom,
            md: md,
            urng: urng,
            cfg: &cfg,
        };

        let (tv, mask) = sba.urng.sample_and_mask().unwrap();
        let (f, _) = sba.do_iteration_tv(0, tv, mask).unwrap();
        let result = fs::read_to_string(f).unwrap();
        assert_display_snapshot!(result);
    }

    #[test]
    fn bv_replace() {
        let fmt_str = Script::Commands(get_bav_assign_fmt_str(&vec![(
            "BAV1".to_owned(),
            Sort::Bool(),
        )]))
        .to_string();

        let bv = BitVec::from_elem(1, true);
        assert_eq!(
            dyn_fmt(&fmt_str, to_strs(&bv)).unwrap().trim(),
            "(assert (= BAV1 true))"
        );
    }
}
