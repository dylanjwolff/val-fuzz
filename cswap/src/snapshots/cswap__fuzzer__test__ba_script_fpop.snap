---
source: cswap/src/fuzzer.rs
expression: new_script
---
(declare-const BAV1 Bool)
(declare-const BDOM1 Bool)
(declare-const BDOM2 Bool)
(declare-const BDOM3 Bool)
(declare-const BDOM4 Bool)
(declare-const BDOM5 Bool)
(declare-const BDOM6 Bool)
(declare-const BDOM7 Bool)
(declare-const BDOM8 Bool)
(declare-const GEN1 Real)
(declare-const GEN2 Int)
(declare-const GEN3 Real)
(declare-const GEN4 Int)
(declare-const mpfx (_ FloatingPoint 11 53))
(assert (not (= mpfx (fp.mul roundTowardPositive ((_ to_fp 11 53) roundNearestTiesToEven GEN1 (- GEN2)) ((_ to_fp 11 53) roundNearestTiesToEven GEN3 (- GEN4))))))
(assert (= BAV1 (= mpfx (fp.mul roundTowardPositive ((_ to_fp 11 53) roundNearestTiesToEven GEN1 (- GEN2)) ((_ to_fp 11 53) roundNearestTiesToEven GEN3 (- GEN4))))))
(assert (= BDOM1 BAV1))
(assert (= (= mpfx (_ -oo 11 53)) BDOM2))
(assert (= (and (fp.gt mpfx (_ -oo 11 53)) (fp.lt mpfx (_ -zero 11 53))) BDOM3))
(assert (= (= mpfx (_ -zero 11 53)) BDOM4))
(assert (= (= mpfx (_ +zero 11 53)) BDOM5))
(assert (= (and (fp.gt mpfx (_ +zero 11 53)) (fp.lt mpfx (_ +oo 11 53))) BDOM6))
(assert (= (= mpfx (_ +oo 11 53)) BDOM7))
(assert (= (= mpfx (_ NaN 11 53)) BDOM8))

